walletconnectv2/src/main/kotlin/org/walletconnect/walletconnectv2/engine/EngineInteractor.kt

```
@@ -126,185 +146,254 @@ internal class EngineInteractor {
-        tryDeserialize<PostSettlementSession.SessionDelete>(json)?.let { sessionDelete ->
+    private fun onSessionDelete(decryptedMessage: String, topic: Topic) {
+        tryDeserialize<PostSettlementSession.SessionDelete>(decryptedMessage)?.let { sessionDelete ->
             //TODO Add subscriptionId from local storage + Delete all data from local storage coupled with given session
-            crypto.removeKeys(topic.topicValue)
+            val (_, selfPublic) = crypto.getKeyAgreement(topic)
+            storageRepository.delete(selfPublic.keyAsHex)
+            crypto.removeKeys(topic.value)
             relayRepository.unsubscribe(topic, SubscriptionId("1"))
             val reason = sessionDelete.message
-            _sequenceEvent.value = SequenceLifecycleEvent.OnSessionDeleted(EngineData.DeletedSession(topic.topicValue, reason))
+            _sequenceEvent.value = SequenceLifecycleEvent.OnSessionDeleted(EngineData.DeletedSession(topic.value, reason))
         } ?: throw NoSessionDeletePayloadException()
     }
 
+    private fun onPairingDelete(json: String, topic: Topic) {
+        //TODO Add subscriptionId from local storage + Delete all data from local storage coupled with given Pairing
+        crypto.removeKeys(topic.value)
+        relayRepository.unsubscribe(topic, SubscriptionId("1"))
+    }
+
+    private fun onPing(topic: Topic, requestId: Long) {
+        val jsonRpcResult = EngineData.JsonRpcResponse.JsonRpcResult(id = requestId, result = "true")
+        val json = trySerialize(jsonRpcResult)
+        val (sharedKey, selfPublic) = crypto.getKeyAgreement(topic)
+        val encryptedMessage: String = codec.encrypt(json, sharedKey as SharedKey, selfPublic as PublicKey)
+        relayRepository.publish(topic, encryptedMessage) { result ->
+            result.fold(
+                onSuccess = {},
+                onFailure = { error -> Logger.error("Ping Error: $error") })
+        }
+    }
+
```

Every time, that we the session is settled and we're gonna use it, we should check if under given topic session exists. If not, we should throw an error. 

Not sure if it's the scope of current PR, let me know.

---

I don't understand what we could do here? Please create an issue and we can address it along with the other errors that we need to handle